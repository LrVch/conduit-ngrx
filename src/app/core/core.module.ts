import { NgModule, Optional, SkipSelf } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { HttpTokenInterceptor } from './interceptors/http.token.interceptor';

import {
  ApiService,
  ArticlesService,
  AuthGuard,
  CommentsService,
  JwtService,
  ProfilesService,
  TagsService,
  UserService,
  CustomEventManager,
  ScrollService,
  DomUtilService
} from './services';
import { EventManager } from '@angular/platform-browser';

@NgModule({
  imports: [
    CommonModule
  ],
  providers: [
    { provide: HTTP_INTERCEPTORS, useClass: HttpTokenInterceptor, multi: true },
    ApiService,
    ArticlesService,
    AuthGuard,
    CommentsService,
    JwtService,
    ProfilesService,
    TagsService,
    UserService,
    ScrollService,
    DomUtilService,
    { provide: EventManager, useClass: CustomEventManager }
  ],
  declarations: []
})
export class CoreModule {
  constructor (@Optional() @SkipSelf() parentModule: CoreModule) {
    if (parentModule) {
      throw new Error(
        'CoreModule is already loaded. Import it in the AppModule only');
    }
  }
 }
