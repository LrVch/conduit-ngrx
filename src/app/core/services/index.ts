export * from './api.service';
export * from './articles.service';
export * from './auth-guard.service';
export * from './comments.service';
export * from './profiles.service';
export * from './tags.service';
export * from './user.service';
export * from './custom-event-manager';
export * from './main-loader.service';
export * from './scroll.service';
export * from './dom-util.service';
export * from './local-storage.service';
export * from './animations.service';
