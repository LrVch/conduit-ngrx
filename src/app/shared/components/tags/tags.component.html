<app-dummy-tags *ngIf="isLoading"></app-dummy-tags>
<div class="tags">
  <mat-chip-list>
    <mat-chip (click)="onSelectTag(tag, currentTag)"
              [selected]="currentTag === tag"
              [color]="currentTag === tag ? 'accent' : ''"
              *ngFor="let tag of tags">{{ tag }}</mat-chip>
  </mat-chip-list>
  <div style="color: red;"
       *ngIf="isErrorLoading">error loading tags</div>
</div>