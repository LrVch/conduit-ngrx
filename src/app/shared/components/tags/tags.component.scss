.tags {
  display: flex;
  justify-content: center;

  mat-chip {
    cursor: pointer;
  }
}
