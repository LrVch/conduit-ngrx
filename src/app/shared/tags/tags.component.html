<div class="tags">
  <mat-chip-list>
    <mat-chip (click)="onSelectTag(tag, currentTag)"
              [selected]="currentTag === tag"
              [color]="currentTag === tag ? 'accent' : ''"
              *ngFor="let tag of tags">{{ tag }}</mat-chip>
  </mat-chip-list>
  <mat-spinner *ngIf="isLoading"></mat-spinner>
  <div style="color: red;"
       *ngIf="isErrorLoading">error loading tags</div>
</div>