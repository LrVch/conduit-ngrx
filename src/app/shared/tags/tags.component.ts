import { Component, Input, Output, EventEmitter } from '@angular/core';


@Component({
  selector: 'app-tags',
  templateUrl: './tags.component.html',
  styleUrls: ['./tags.component.scss']
})
export class TagsComponent {
  @Input('tags') tags: string[] = [];
  @Input('currentTag') currentTag = '';
  @Input('isLoading') isLoading = false;
  @Input('isErrorLoading') isErrorLoading = false;
  @Output() selecteTag = new EventEmitter();

  onSelectTag(tag: string, currentTag: string): void {
    if (tag === currentTag) {
      return;
    }

    this.selecteTag.emit(tag);
  }
}
