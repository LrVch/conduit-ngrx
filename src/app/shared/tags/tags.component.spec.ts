import { async, ComponentFixture } from '@angular/core/testing';

import { ConfigureFn, configureTests } from '../../lib/testing';

import { TagsComponent } from './tags.component';
import { NO_ERRORS_SCHEMA, DebugElement } from '@angular/core';
import { By } from '@angular/platform-browser';

describe('TagsComponent', () => {
    let component: TagsComponent;
    let fixture: ComponentFixture<TagsComponent>;
    let de: DebugElement;
    let el: HTMLElement;

    beforeEach(
        async(() => {
            const configure: ConfigureFn = testBed => {
                testBed.configureTestingModule({
                    declarations: [TagsComponent],
                    schemas: [NO_ERRORS_SCHEMA]
                });
            };

            configureTests(configure).then(testBed => {
                fixture = testBed.createComponent(TagsComponent);
                component = fixture.componentInstance;
                de = fixture.debugElement;
                el = de.nativeElement;
                fixture.detectChanges();
            });
        })
    );

    it('should create', () => {
        expect(component).toBeTruthy();
    });

    it('should has "tags" @Input', () => {
        expect(component.tags).toEqual([]);
    });

    it('should has "currentTag" @Input', () => {
        expect(component.currentTag).toBe('');
    });

    it('should has "isLoading" @Input', () => {
        expect(component.isLoading).toBeFalsy();
    });

    it('should has "isErrorLoading" @Input', () => {
        expect(component.isErrorLoading).toBeFalsy();
    });

    it('should show "tags" @Input', () => {
        component.tags = ['one', 'two', 'three'];
        fixture.detectChanges();
        expect(fixture).toMatchSnapshot();

        component.tags = ['one'];
        fixture.detectChanges();
        expect(fixture).toMatchSnapshot();

        component.tags = [''];
        fixture.detectChanges();
        expect(fixture).toMatchSnapshot();

        component.tags = [];
        fixture.detectChanges();
        expect(fixture).toMatchSnapshot();
    });

    it('should mark tag as selected via "currentTag" @Input and set color of tag to "accent"', () => {
        component.currentTag = 'one';
        component.tags = ['one', 'two', 'three'];

        fixture.detectChanges();

        const firstTagElem = de.query(By.css('mat-chip')).nativeElement;

        expect(firstTagElem.selected).toBeTruthy();
        expect(firstTagElem.color).toBe('accent');
    });

    it('should not raise "selecteTag" event if clicked tag is equals to currentTag', () => {
        let wasEmited = false;

        component.currentTag = 'one';
        component.tags = ['one', 'two', 'three'];
        component.selecteTag.subscribe((tag: string) => {
            console.log(tag);
            wasEmited = !!tag;
        });

        fixture.detectChanges();

        const firstTag = de.query(By.css('mat-chip'));

        firstTag.triggerEventHandler('click', null);

        expect(wasEmited).toBeFalsy();
    });

    it('should show loader', () => {
        component.isLoading = true;

        fixture.detectChanges();
        expect(fixture).toMatchSnapshot();
    });

    it('should show error', () => {
        component.isErrorLoading = true;

        fixture.detectChanges();

        const errorMessElem = de.query(By.css('[style]')).nativeElement;

        expect(fixture).toMatchSnapshot();
        expect(errorMessElem.textContent).toContain('error');
    });

    it('should raise selected event when clicked (selecteTag) and select first tag', () => {
        let selectedTag: string;
        const expectedTag = 'one';
        component.tags = ['one', 'two', 'three'];
        component.selecteTag.subscribe((tag: string) => selectedTag = tag);

        fixture.detectChanges();

        const firstTag = de.query(By.css('mat-chip'));

        firstTag.triggerEventHandler('click', null);
        expect(selectedTag).toBe(expectedTag);
    });
});

