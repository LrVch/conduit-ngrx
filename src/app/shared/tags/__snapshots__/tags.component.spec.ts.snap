// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagsComponent should mark tag as selected via "currentTag" @Input and set color of tag to "accent" 1`] = `
<app-tags
  currentTag={[Function String]}
  isErrorLoading="false"
  isLoading="false"
  selecteTag={[Function EventEmitter]}
  tags={[Function Array]}
>
  <div
    class="tags"
  >
    <mat-chip-list
      aria-disabled="false"
      aria-invalid="false"
      aria-multiselectable="false"
      aria-orientation="horizontal"
      aria-required="false"
      class="mat-chip-list"
      id="mat-chip-list-6"
      role="listbox"
      tabindex="0"
    >
      <div
        class="mat-chip-list-wrapper"
      >
        
        <mat-chip
          aria-disabled="false"
          aria-selected="true"
          class="mat-chip mat-standard-chip mat-accent mat-chip-selected"
          ng-reflect-color="accent"
          ng-reflect-selected="true"
          role="option"
          tabindex="-1"
        >
          one
        </mat-chip>
        <mat-chip
          aria-disabled="false"
          aria-selected="false"
          class="mat-chip mat-primary mat-standard-chip"
          ng-reflect-color=""
          ng-reflect-selected="false"
          role="option"
          tabindex="-1"
        >
          two
        </mat-chip>
        <mat-chip
          aria-disabled="false"
          aria-selected="false"
          class="mat-chip mat-primary mat-standard-chip"
          ng-reflect-color=""
          ng-reflect-selected="false"
          role="option"
          tabindex="-1"
        >
          three
        </mat-chip>
      </div>
    </mat-chip-list>
    
    
  </div>
</app-tags>
`;

exports[`TagsComponent should show "tags" @Input 1`] = `
<app-tags
  currentTag=""
  isErrorLoading="false"
  isLoading="false"
  selecteTag={[Function EventEmitter]}
  tags={[Function Array]}
>
  <div
    class="tags"
  >
    <mat-chip-list
      aria-disabled="false"
      aria-invalid="false"
      aria-multiselectable="false"
      aria-orientation="horizontal"
      aria-required="false"
      class="mat-chip-list"
      id="mat-chip-list-5"
      role="listbox"
      tabindex="0"
    >
      <div
        class="mat-chip-list-wrapper"
      >
        
        <mat-chip
          aria-disabled="false"
          aria-selected="false"
          class="mat-chip mat-primary mat-standard-chip"
          ng-reflect-color=""
          ng-reflect-selected="false"
          role="option"
          tabindex="-1"
        >
          one
        </mat-chip>
        <mat-chip
          aria-disabled="false"
          aria-selected="false"
          class="mat-chip mat-primary mat-standard-chip"
          ng-reflect-color=""
          ng-reflect-selected="false"
          role="option"
          tabindex="-1"
        >
          two
        </mat-chip>
        <mat-chip
          aria-disabled="false"
          aria-selected="false"
          class="mat-chip mat-primary mat-standard-chip"
          ng-reflect-color=""
          ng-reflect-selected="false"
          role="option"
          tabindex="-1"
        >
          three
        </mat-chip>
      </div>
    </mat-chip-list>
    
    
  </div>
</app-tags>
`;

exports[`TagsComponent should show "tags" @Input 2`] = `
<app-tags
  currentTag=""
  isErrorLoading="false"
  isLoading="false"
  selecteTag={[Function EventEmitter]}
  tags={[Function Array]}
>
  <div
    class="tags"
  >
    <mat-chip-list
      aria-disabled="false"
      aria-invalid="false"
      aria-multiselectable="false"
      aria-orientation="horizontal"
      aria-required="false"
      class="mat-chip-list"
      id="mat-chip-list-5"
      role="listbox"
      tabindex="0"
    >
      <div
        class="mat-chip-list-wrapper"
      >
        
        <mat-chip
          aria-disabled="false"
          aria-selected="false"
          class="mat-chip mat-primary mat-standard-chip"
          ng-reflect-color=""
          ng-reflect-selected="false"
          role="option"
          tabindex="-1"
        >
          one
        </mat-chip>
      </div>
    </mat-chip-list>
    
    
  </div>
</app-tags>
`;

exports[`TagsComponent should show "tags" @Input 3`] = `
<app-tags
  currentTag=""
  isErrorLoading="false"
  isLoading="false"
  selecteTag={[Function EventEmitter]}
  tags={[Function Array]}
>
  <div
    class="tags"
  >
    <mat-chip-list
      aria-disabled="false"
      aria-invalid="false"
      aria-multiselectable="false"
      aria-orientation="horizontal"
      aria-required="false"
      class="mat-chip-list"
      id="mat-chip-list-5"
      role="listbox"
      tabindex="0"
    >
      <div
        class="mat-chip-list-wrapper"
      >
        
        <mat-chip
          aria-disabled="false"
          aria-selected="true"
          class="mat-chip mat-standard-chip mat-accent mat-chip-selected"
          ng-reflect-color="accent"
          ng-reflect-selected="true"
          role="option"
          tabindex="-1"
        >
          
        </mat-chip>
      </div>
    </mat-chip-list>
    
    
  </div>
</app-tags>
`;

exports[`TagsComponent should show "tags" @Input 4`] = `
<app-tags
  currentTag=""
  isErrorLoading="false"
  isLoading="false"
  selecteTag={[Function EventEmitter]}
  tags={[Function Array]}
>
  <div
    class="tags"
  >
    <mat-chip-list
      aria-disabled="false"
      aria-invalid="false"
      aria-multiselectable="false"
      aria-orientation="horizontal"
      aria-required="false"
      class="mat-chip-list"
      id="mat-chip-list-5"
      tabindex="-1"
    >
      <div
        class="mat-chip-list-wrapper"
      >
        
      </div>
    </mat-chip-list>
    
    
  </div>
</app-tags>
`;

exports[`TagsComponent should show error 1`] = `
<app-tags
  currentTag=""
  isErrorLoading={[Function Boolean]}
  isLoading="false"
  selecteTag={[Function EventEmitter]}
  tags={[Function Array]}
>
  <div
    class="tags"
  >
    <mat-chip-list
      aria-disabled="false"
      aria-invalid="false"
      aria-multiselectable="false"
      aria-orientation="horizontal"
      aria-required="false"
      class="mat-chip-list"
      id="mat-chip-list-9"
      tabindex="-1"
    >
      <div
        class="mat-chip-list-wrapper"
      >
        
      </div>
    </mat-chip-list>
    
    
    <div
      style="color: red;"
    >
      error loading tags
    </div>
  </div>
</app-tags>
`;

exports[`TagsComponent should show loader 1`] = `
<app-tags
  currentTag=""
  isErrorLoading="false"
  isLoading={[Function Boolean]}
  selecteTag={[Function EventEmitter]}
  tags={[Function Array]}
>
  <div
    class="tags"
  >
    <mat-chip-list
      aria-disabled="false"
      aria-invalid="false"
      aria-multiselectable="false"
      aria-orientation="horizontal"
      aria-required="false"
      class="mat-chip-list"
      id="mat-chip-list-8"
      tabindex="-1"
    >
      <div
        class="mat-chip-list-wrapper"
      >
        
      </div>
    </mat-chip-list>
    
    <mat-spinner
      class="mat-spinner mat-progress-spinner mat-primary mat-progress-spinner-indeterminate-animation"
      mode="indeterminate"
      role="progressbar"
      style="width: 100px; height: 100px;"
    >
      <svg
        focusable="false"
        ng-reflect-ng-switch="true"
        preserveAspectRatio="xMidYMid meet"
        style="width: 100px; height: 100px;"
        viewBox="0 0 100 100"
      >
        
        <circle
          cx="50%"
          cy="50%"
          r="45"
          style="stroke-dasharray: 282.7433388230814px; stroke-width: 10%;"
        />
        
      </svg>
    </mat-spinner>
    
  </div>
</app-tags>
`;
